common_arm/flashmem.c

```diff
@@ -43,7 +43,34 @@ static uint32_t FLASHMEM_SPIBAUDRATE = FLASH_BAUD;
 
 #ifndef AS_BOOTROM
 
-uint8_t spi_flash_p64k = 0;
+const spi_flash_t SpiFlashTable[] = {
+    // first element is the default of 4 * 64kB pages (256kB)
+    { 0x00, 0x00, 0x0000,  4, "unknown" },     //  256k
+    // Manufacturer: Puya
+    { 0x85, 0x00, 0x6015, 32, "P25Q16H" },     // 2048k
+    /// Manufacturer: Renesas
+    { 0x1F, 0x46, 0x0000, 32, "AT25XE161D" },  // 2048k
+    { 0x1F, 0x47, 0x0000, 64, "AT25XE321D" },  // 4096k
+    // Manufacturer: Winbond
+    { 0xEF, 0x00, 0x3012,  4, "W25X20BV" },    //  256k
+    { 0xEF, 0x00, 0x3013,  8, "W25X40BV" },    //  512k
+
+    { 0xEF, 0x00, 0x4013,  8, "W25Q40BV" },    //  512k
+    { 0xEF, 0x00, 0x4014, 16, "W25Q80BV" },    // 1024k
+    { 0xEF, 0x14, 0x4015, 32, "W25Q16BV" },    // 2048k
+    { 0xEF, 0x15, 0x4016, 64, "W25Q32BV" },    // 4096k
+
+    { 0xEF, 0x21, 0x7022,  4, "W25Q02JV" },
+    // identified by Manufacturer /Device ID
+//    { 0xEF, 0x05, 0x0000,  1, "Winbond!!!" },
+    { 0xEF, 0x10, 0x0000,  2, "W25*10BV!!!" }, //  128k
+    { 0xEF, 0x11, 0x0000,  4, "W25*20BV" },    //  256k
+    { 0xEF, 0x12, 0x0000,  8, "W25*40BV" },    //  512k
+    { 0xEF, 0x13, 0x0000, 16, "W25*80BV" }     // 1024k
+};
+const spi_flash_t *spi_flash_p = SpiFlashTable;
```

Comment:
If you move this table to the header file where you declare it,  you can remove the extern reference to it. 
Not sure why you complicate it with pointers.  I would have a static extern for table_idx,   and a "get_item" function.

Reply:
You are right. 
Actually only the number of the 64k pages is needed outside of the flashmem.c file.
rework in progress...